[test][sagemaker] Parallelize all SageMaker Integration remote tests

saimidu opened the pull request:
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
- Add `-n=logical` to SM Remote test pytest commands
- Upgrade pytest and associated package versions on all requirements.txt files
- Enable "retry until time-limit" on all sm test helper functions
- Use sm test helper functions with retry capabilities on all tests

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests for my changes.
    - [998b238](https://github.com/aws/deep-learning-containers/pull/3102/commits/998b238c9dad8f8109b40cbdcdaf66b924736e58) for PyTorch and TensorFlow DLCs
    - [998b238](https://github.com/aws/deep-learning-containers/pull/3102/commits/998b238c9dad8f8109b40cbdcdaf66b924736e58) for AutoGluon Training DLCs
    - [895352b](https://github.com/aws/deep-learning-containers/pull/3102/commits/895352b361bbdb5ace0cbb9ab17b0e857e45eb5b) for AutoGluon Inference DLCs
    - [21c9617](https://github.com/aws/deep-learning-containers/pull/3102/commits/21c9617c0f054b4b10e59d32a21fe61541a837cc) for Hugging Face PT and TF DLCs
    - [e3244c4](https://github.com/aws/deep-learning-containers/pull/3102/commits/e3244c4a5b87b0b9423a211511cfb7ace982daa0) for Hugging Face PT and TF DLCs after resolving merge conflicts in HF PT tests
    - [d2102a8](https://github.com/aws/deep-learning-containers/pull/3102/commits/d2102a87d6d2c0b3c269eebfe7f9d8f1eb9dd691) for RC tests on PT, TF, HF-PT, HF-TF
    - [edec708](https://github.com/aws/deep-learning-containers/pull/3102/commits/edec70854555ddf4f9220449b2e90450dc68c6db) for EFA tests on PT, TF, HF-PT, HF-TF

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [x] Revision B: `sagemaker_remote_tests = "rc"`
- [x] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

### DLC image/dockerfile

### Additional context

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

saimidu commented:
All TF SM Training tests succeed at commit [69f1737](https://github.com/aws/deep-learning-containers/pull/3102/commits/69f173701ad35785c946f54fe5fe115db15129d3) without needing to be restarted to handle EC2 Instance `CapacityErrors` and SM Client `ThrottlingExceptions`.

saimidu commented:
All TF SM Inference tests succeed at commit [029fb43](https://github.com/aws/deep-learning-containers/pull/3102/commits/029fb43d68dfc9dd6818aaf24eccd778cc2cf60a).

saimidu commented:
All PT SM Training and Inference tests succeed at commit [eccdd02](https://github.com/aws/deep-learning-containers/pull/3102/commits/eccdd020fc364228be1effd4e2cdbc6500713046) without needing to be restarted to handle EC2 Instance `CapacityErrors` and SM Client `ThrottlingExceptions`.

saimidu commented on huggingface/pytorch/inference/buildspec.yml:
```diff
@@ -27,19 +27,19 @@ context:
-    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile., *DEVICE_TYPE ]
-    context:
-      <<: *INFERENCE_CONTEXT
+#  BuildHuggingFacePytorchCpuPy310InferenceDockerImage:
+#    <<: *INFERENCE_REPOSITORY
+#    build: &HUGGINGFACE_PYTORCH_CPU_INFERENCE_PY3 false
+#    image_size_baseline: 15000
+#    device_type: &DEVICE_TYPE cpu
+#    python_version: &DOCKER_PYTHON_VERSION py3
+#    tag_python_version: &TAG_PYTHON_VERSION py310
+#    os_version: &OS_VERSION ubuntu20.04
+#    transformers_version: &TRANSFORMERS_VERSION 4.28.1
+#    tag: !join [ *VERSION, '-', 'transformers', *TRANSFORMERS_VERSION, '-', *DEVICE_TYPE, '-', *TAG_PYTHON_VERSION, '-', *OS_VERSION ]
+#    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile., *DEVICE_TYPE ]
+#    context:
+#      <<: *INFERENCE_CONTEXT
```
Skipping builds on HF PT CPU DLC because of ipex installation errors.

saimidu commented:
All HF TF and HF PT Training and Inference tests succeed at commit [82f8650](https://github.com/aws/deep-learning-containers/pull/3102/commits/82f8650b39eeb33383540a7aa31570a783e59b0b).

HF PT Training SMMP tests still take ~30-40 mins due to the actual training job being run, but overall test time is reduced significantly either way.

saimidu commented:
All SM Training and Inference tests pass for Graviton, Neuron, and NeuronX DLCs except for build-jobs that have dummy buildspecs as images do not exist.

xncqr commented:
In terms of Graviton, I think we are good with the changes as they are. Thank you for updating pytest.

shantanutrip commented on test/sagemaker_tests/pytorch/training/integration/sagemaker/test_mnist_inductor.py:
```diff
@@ -29,7 +29,7 @@
 @pytest.mark.skip_gpu
 @pytest.mark.skip_inductor_test
-@pytest.mark.xfail(reason="known issue: https://github.com/pytorch/pytorch/issues/98436")
+@pytest.mark.skip(reason="known issue: https://github.com/pytorch/pytorch/issues/98436")
```
Do we want to change from `xfail` to `skip`? In the future, in case the issue is resolved and the tests start to pass, if we use `skip`, we would not be able to catch the passing tests and would not be able to rerun the tests. 

In other words, with `skip` we might never get to run the tests even when the issue is resolved.

saimidu replied:
If the issue is resolved in the future, we'll need to unskip these tests.

As of now, due to the use of `xfail`, the overall behavior of the test job is exactly the same regardless of whether these inductor tests pass or fail, but we waste valuable time and money on running a test that we know is going to fail.

I've created a GitHub Issue to track this pending task - https://github.com/aws/deep-learning-containers/issues/3128

Right now, until we remove the xfail marker, we'll never know if the test is working correctly anyway.

shantanutrip commented on test/sagemaker_tests/pytorch/training/requirements.txt:
```diff
@@ -2,7 +2,7 @@ botocore>1.0,<2.0
 boto3>1.0,<2.0
 awscli>=1.27.51
 protobuf
-sagemaker==2.152.0
```
I guess there was a bug in one of the SM versions, because of which we had to fix this version. In case the tests are passing, I think it is fine to remove the pin from SM.

saimidu replied:
Yeah, we had to pin sagemaker, fabric + invoke, and other packages due to bugs that existed for a short period of time. Upon discovering that removing the version-pin doesn't break the tests on one of the versions, I unpinned them all, because older versions have dependencies that prevent the use of the latest pytest and pytest optimizations.

shantanutrip commented on test/sagemaker_tests/__init__.py:
```diff
@@ -115,43 +145,105 @@ def invoke_sm_helper_function(ecr_image, sagemaker_regions, test_function, *test
-        )
+
+        # Run model setup if needed
+        tested_model_data = None
+        if model_helper and local_model_paths:
+            if not mme_folder_name:
+                tested_model_data = model_helper(
+                    region=region, boto_session=boto_session, local_path=local_model_paths[0]
+                )
+            else:
+                tested_model_data = model_helper(
+                    region=region,
+                    boto_session=boto_session,
+                    mme_folder_name=mme_folder_name,
+                    path_list=local_model_paths,
+                )
```
While reading the entire PR, I realized that the reason could be that this method needs to be much more generic now since it is being used by various tests.

saimidu replied:
Yeah, the behavior of `model_helper` is different for MME tests and regular TFS endpoint tests.

And in non-TF inference tests, this whole model_helper section is completely unnecessary. Since I'm reusing this function to run SM inference tests for all frameworks, I've tried to make it capable of handling all scenarios.

shantanutrip commented on test/test_utils/sagemaker.py:
```diff
@@ -226,9 +226,11 @@ def generate_sagemaker_pytest_cmd(image, sagemaker_test_type):
-        f"pytest -rA {integration_path} --region {region} --processor {processor} {docker_base_arg} "
-        f"{sm_remote_docker_base_name} --tag {tag} {framework_version_arg} {framework_version} "
-        f"{aws_id_arg} {account_id} {instance_type_arg} {instance_type} {efa_flag} {sagemaker_regions_list} --junitxml {test_report}"
+        f"pytest -n=logical --dist worksteal -rA {integration_path} --region {region} "
```
Can we add some comments, especially to explain `-n=logical` and `--dist`, to ensure that contributors understand the importance of the commands?

saimidu replied:
Sure, I'll add a comment. I checked to see whether we had comments in other places for pytest args, and decided not to add explanatory comments because that was less work.

However explaining why we're adding certain pytest args for each type of test would inform engineers making any further changes in the future.

Explaining it here as well -
- auto vs logical, and test distribution algorithms - https://pytest-xdist.readthedocs.io/en/stable/distribution.html
- `-n=logical` does nothing on CI/CD, because CodeBuild already provides single-threaded vCPUs at the OS-level to jobs, which means that any parallelization already splits work into one worker per vCPU when using `-n=auto`. However, using `-n=auto` outside CodeBuild (such as on a personal computer or EC2 instance) spawns `num_cpu_cores` parallel worker processes, while using `-n=logical` spawns `num_cpu_cores * 2` worker processes.
- The default `--dist=load` option splits all tests into queues between the available workers, and then lets them handle their own workload. `--dist=worksteal` allows any processes that finish their tests faster to steal tests from the queues for slower workers. This is helpful when our automation needs to run, say, 2 tests while skipping 8 other tests, and we want to have the 2 tests actually be split between workers instead of running the 2 test sequentially because they were scheduled onto the same worker process queue.

shantanutrip reviewed (commented):
The PR looks good. The comments that say `Minor change` are simple suggestions and it should be fine to not add the fix for those in this PR. Thanks!!

saimidu commented on test/sagemaker_tests/__init__.py:
```diff
@@ -115,43 +145,105 @@ def invoke_sm_helper_function(ecr_image, sagemaker_regions, test_function, *test
-
-
+        except botocore.exceptions.ClientError as e:
+            if "ThrottlingException" in str(e) or "ResourceLimitExceeded" in str(e):
```
Good catch. Will make the change. We may encounter more than just these 2 types of errors to add exceptions for in the future.

saimidu commented on test/sagemaker_tests/__init__.py:
```diff
@@ -115,43 +145,105 @@ def invoke_sm_helper_function(ecr_image, sagemaker_regions, test_function, *test
+        except Exception as e:
+            if dump_logs_from_cloudwatch:
+                dump_logs_from_cloudwatch(e, region)
+            raise
```
`raise` works the same as `raise e`.

saimidu commented on test/sagemaker_tests/autogluon/inference/integration/sagemaker/test_serving.py:
```diff
@@ -26,10 +26,13 @@
 @contextmanager
-def _test_sm_trained_model(sagemaker_session, ecr_image, instance_type, framework_version):
+def _test_sm_trained_model(
+    sagemaker_session, image_uri, instance_type, framework_version, **kwargs
```
I changed it to `image_uri` in order to make it more similar to the SageMaker Python SDK API parameter name for image URI.

I had to choose whether to change helper functions for all tests to use one of `ecr_image` or `image_uri`, to pass the image URI as a "regionalized" keyword-arg to the helper. So I chose `image_uri`.

saimidu commented on test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_diffusers_model.py:
```diff
@@ -24,64 +26,68 @@
 ):
+    framework, _ = get_framework_and_version_from_tag(ecr_image)
+    if "pytorch" not in framework:
+        pytest.skip(f"Skipping test for non-pytorch image - {ecr_image}")
```
The code was written to raise an AssertionError on non PyTorch DLC images, which caused this HF-Inference test to fail on HF-TF Inference tests. So I changed it to skip the test when encountering a non-PT image.

tejaschumbalkar reviewed (commented):
LGTM

saimidu commented:
Unable to root cause the coverage test failure, but it is likely caused by local import failures due to files/folders named `sagemaker` being detected in the import path when running the coverage test in such a way that doesn't occur during normal testing.

Removing `import sagemaker.exceptions` and reverting imports to their previous state doesn't break anything, it just creates annoying warnings on import-checks on IDEs, but does not affect actual code functionality.

saimidu commented:
Coverage test issues resolved in the short term by removing imports of specific submodules/classes from the sagemaker package.

Rerunning all tests once more to ensure that nothing broke.

saimidu commented on test/sagemaker_tests/autogluon/inference/requirements.txt:
```diff
@@ -1,25 +1,20 @@
 boto3
```
Thanks!

saimidu commented on test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_ipex_inference.py:
```diff
@@ -25,39 +25,40 @@
-):
+def test_ipex_hosting(framework_version, ecr_image, instance_type, sagemaker_regions, py_version):
+    framework, _ = get_framework_and_version_from_tag(ecr_image)
+    if "pytorch" not in framework:
```
Thanks!

saimidu commented:
Made the following changes since the last review:
1. Addressed comments from @shantanutrip's review
2. Fixed quick-check coverage test failures (changed some import statements in the sagemaker retry-helper functions)
3. Re-ran all SM tests and discovered bug in one of the AG SM Inference test where framework_version wasn’t passed in from the retry-helper function
4. Resolved a merge conflict with the latest Hugging Face PT 1.13 PR merge
5. Re-ran all SM Training and Inference tests on all frameworks
6. Re-ran SM tests in RC test mode
7. Re-ran SM tests in EFA test mode
8. Reverted all toml file and buildspec changes

shantanutrip commented on test/test_utils/sagemaker.py:
```diff
@@ -225,10 +225,20 @@ def generate_sagemaker_pytest_cmd(image, sagemaker_test_type):
 
+    # Auto vs logical, and test distribution algorithms -
+    #   https://pytest-xdist.readthedocs.io/en/stable/distribution.html
+    # Using -n=logical spawns num_cpu_cores * 2 worker processes, instead of -n=auto which spawns
```
Thanks for adding this!
